### ISSUE TITLE
Read aspect providers in cquery

### ISSUE THREAD

**Issue by matts1** [bazelbuild/bazel#22528]
### Description of the feature request:

Cquery should be able to read providers that were created by an aspect.

I tested, and the aspects already run during cquery (see https://github.com/bazelbuild/bazel/issues/16310#issuecomment-2113976408). So all I'm asking is that we surface that via cquery. I suspect that this will *likely* be a relatively simple change, since the aspects already run, but it depends on how it's implemented.

### Which category does this issue belong to?

Rules API

### What underlying problem are you trying to solve with this feature?

Aspects are very useful for allowing IDEs access to metadata about the build. However, at the moment, IIUC, the way you have to use them is to:
1) run `bazel build <something> --aspects=...`
1) Read from the output file generated by your aspect.

Sometimes, however, you don't need anything other than the providers created by the aspect. In cases like this, cquery should be sufficient. For example, generating `rust-project.json` in rust can be done without building a single target.

The use case that I was trying to do for this is that if I run cquery on the following target:
```
cc_binary(
    name = "foo",
    deps = ["//path/to:rust_library"],
)
```

Then I want cquery to attach a `RustAnalyzerInfo` object to `:foo`. This can then be picked up on by the rust-analyzer vscode extension. At the moment we use `bazel build` instead of `bazel cquery`, which means that if *any* bazel target fails to build, then we can't run get IDE support for our rust code.

### Which operating system are you running Bazel on?

Linux

### What is the output of `bazel info release`?

7.1.1

### Have you found anything relevant by searching the web?

I commented on https://github.com/bazelbuild/bazel/issues/16310#issuecomment-2113976408 last week

**Comment by mai93** [comment]
Can specifying *only* the output containing the file with aspect provider data via [`--output_groups`](https://bazel.build/reference/command-line-reference#flag--output_groups) in the `bazel build <something> --aspects=...` invocation help with that? IIUC this will not execute the rest of the actions in the build that may fail.

**Comment by matts1** [comment]
That's a good point. I just tried it out, and it does appear to work. I still think this suggestion has value, but it can be achieved via other means, so it's probably not a priority.

**Comment by kilian-funk** [comment]
The `bazel build <something> --aspect=...` workaround is only practical for a few targets. It is not helpful if I want to use this approach to harvest more detailed information on a complete region of my workspace. In my use case, I need to process certain information from an attribute of many targets. Since cquery Starlark does not give me access to the attributes, I need to use an aspect. Using output groups in combination with bazel build would mean reading thousands of files to collect the results.

So I would be very happy if this issue got fixed soon :+1:

**Comment by fmeum** [comment]
Not stale. I worked on this for a bit, but noticed that a male problem is how to properly present the result. 

A given target can be visited by multiple aspects which can even apply to each other in a DAG and, at least in theory, each aspect could return a different value for a given provider. A single merged view doesn't exist. Perhaps we need an `aspects` dict similar to `providers`?

Happy to work on this if someone comes up with a decent API.

**Comment by stagnation** [comment]
Good point. Yes a dictionary would serve most of my purposes. Don't have a use-case at hand but I remember being confused by this and have many little cquery scripts for printing the providers. Those would be more useful if they also showed providers created by aspects, and I could easily iterate through a dictionary to list all the aspects that have been applied and the extra providers they create.
